Implement a proper sparsity pattern (#888)
Ferrite-FEM/Ferrite.jl: docs/src/topics/sparse_matrix.md

are discussed.

```@contents
Pages = ["sparse_matrix.md"]
Depth = 2:2
```

## Sparsity pattern

The sparse structure of the linear system depends on many factors such as e.g. the weak
form, the discretization, and the choice of interpolation(s). In the end it boils down to
how the degrees of freedom (DoFs) *couple* with each other. The most common reason that two
DoFs couple is because they belong to the same element. Note, however, that this is not
guaranteed to result in a coupling since it depends on the specific weak form that is being
discretized, see e.g. [Increasing the sparsity](@ref). Boundary conditions and constraints
can also result in additional DoF couplings.

If DoFs `i` and `j` couple, then the computed value in the eventual matrix will be
*structurally nonzero*[^1]. In this case the entry `(i, j)` should be included in the
sparsity pattern. Conversely, if DoFs `i` and `j` *don't* couple, then the computed value
will be *zero*. In this case the entry `(i, j)` should *not* be included in the sparsity
pattern since there is no need to allocate memory for entries that will be zero.

> The sparsity, i.e. the ratio of zero-entries to the total number of entries, is often *very*
> high and taking advantage of this results in huge savings in terms of memory. For example,

Reviewer: Maybe we should be a bit more specific here to not mislead users accidentally.
```suggestion
The sparsity, i.e. the ratio of zero-entries to the total number of entries, is for most practical problems using low-order interpolations *very*
high for  and taking advantage of this results in huge savings in terms of memory. For example,
```
Note that this statement is not true anymore for e.g. high order Ansatz or interpolations spanning multiple elements.

     - Trial function 4 couples with test functions 3 and 4 (entries `(4, 3)` and `(4, 4)`
       included in the sparsity pattern)

    The resulting sparsity pattern would look like this:

    ```
    4×4 SparseArrays.SparseMatrixCSC{Float64, Int64} with 10 stored entries:
     0.0  0.0   ⋅    ⋅
     0.0  0.0  0.0   ⋅
      ⋅   0.0  0.0  0.0
      ⋅    ⋅   0.0  0.0
    ```

    Moreover, if the problem is solved with periodic boundary conditions, for example by
    constraining the value on the right side to the value on the left side, there will be
    additional couplings. In the example above, this means that DoF 4 should be equal to DoF
    1. Since DoF 4 is constrained it has to be eliminated from the system. Existing entries
    that include DoF 4 are `(3, 4)`, `(4, 3)`, and `(4, 4)`. Given the simple constraint in
    this case we can simply replace DoF 4 with DoF 1 in these entries and we end up with
    entries `(3, 1)`, `(1, 3)`, and `(1, 1)`. This results in two new entries: `(3, 1)` and
    `(1, 3)` (entry `(1, 1)` is already included).

## Creating sparsity patterns

Creating a sparsity pattern can be quite expensive if not done properly and therefore

> Ferrite provides efficient methods and datastructures for this. In general the sparsity

Reviewer: ```suggestion
Ferrite provides efficient methods and data structures for this. In general the sparsity
```

pattern is not known in advance and has to be created incrementally. To make this
incremental construction efficient it is necessary to use a dynamic data structure which
allow for fast insertions.

The sparsity pattern also serves as a "matrix builder". When all entries are inserted into
the sparsity pattern the dynamic data structure is typically converted, or "compressed",
into a sparse matrix format such as e.g. the [*compressed sparse row
(CSR)*](https://en.wikipedia.org/wiki/Sparse_matrix#Compressed_sparse_row_(CSR,_CRS_or_Yale_format))
format or the [*compressed sparse column
(CSC)*](https://en.wikipedia.org/wiki/Sparse_matrix#Compressed_sparse_column_(CSC_or_CCS))

> format, where the latter is the default sparse matrix type implemented in the SparseArrays
> standard library. These matrix formats allow for fast linear algebra operations, such as

Reviewer: Maybe we should just add a link to make sure this is clear for new users which are not too familiar with the ecosystem.
```suggestion
format, where the latter is the default sparse matrix type implemented in the [SparseArrays
standard library](https://sparsearrays.juliasparse.org/dev/). These matrix formats allow for fast linear algebra operations, such as
```

       interfaces between elements (e.g. for discontinuous Galerkin methods).
     - [`add_constraint_entries!`](@ref) adds entries required from constraints and boundary
       conditions in the ConstraintHandler. Note that this operation depends on existing
       entries in the pattern and *must* be called as the last operation on the pattern.
     - [`Ferrite.add_entry!`](@ref) adds a single entry to the pattern. This can be used if
       you need to add custom entries that are not covered by the other functions.

 3. **Instantiate the matrix:** A sparse matrix can be created from the sparsity pattern
    using [`allocate_matrix`](@ref), see [Instantiating the sparse matrix](@ref) below for
    more details.

### Increasing the sparsity

By default, when creating a sparsity pattern, it is assumed that each DoF within an element
couple with with *all* other DoFs in the element.

!!! todo
     - Discuss the `coupling` keyword argument.
     - Discuss the `keep_constrained` keyword argument.

### Blocked sparsity pattern

!!! todo
    Discuss `BlockSparsityPattern` and `BlockArrays` extension.

## Instantiating the sparse matrix

!!! todo

>     Write text.

Reviewer: Will this be handled in a separate PR or do you intend to add it here?

# Sparsity pattern and sparse matrix

An important property of the finite element method is that it results in *sparse matrices*
for the linear systems to be solved. On this page the topic of sparsity and sparse matrices
are discussed.

```@contents
Pages = ["sparse_matrix.md"]
Depth = 2:2
```

## Sparsity pattern

The sparse structure of the linear system depends on many factors such as e.g. the weak
form, the discretization, and the choice of interpolation(s). In the end it boils down to
how the degrees of freedom (DoFs) *couple* with each other. The most common reason that two

> DoFs couple is because they simply belong to the same element. Note, however, that this is

Reviewer 2: ```suggestion
DoFs couple is because they belong to the same element. Note, however, that this is
```

not guaranteed to result in a coupling since it depends on the specific weak form that is

> being discretized, see [XXX](@ref). Boundary conditions and constraints can also result in

Reviewer 2: TODO here

additional DoF couplings.

If DoFs `i` and `j` couple, then the computed value in the eventual matrix will be

> *nonzero*. In this case the entry `(i, j)` should be included in the sparsity pattern.

Reviewer 2: Is it worth saying something like "structurally non-zero"? Like, it could happen for various reasons that it becomes zero anyway.

# Sparsity pattern and sparse matrix

An important property of the finite element method is that it results in *sparse matrices*
for the linear systems to be solved. On this page the topic of sparsity and sparse matrices
are discussed.

```@contents
Pages = ["sparse_matrix.md"]
Depth = 2:2
```

## Sparsity pattern

> The sparse structure of the linear system depends on many factors such as e.g. the weak

Reviewer 3: I'm a bit torn here since the derivative is not always the definition of the matrix (but implicitly assumed in the text below for e.g. reference to Poisson's equation). Yet, I find it logical introduce the coupling via the derivative, e.g.

----------
In general, a finite element problem requires solving an equation system
```math
\boldsymbol{r}(\boldsymbol{a}) = 0
```
to find $\boldsymbol{a}$. For nonlinear problems when using the Newton-Raphson method, 
or for linear problems, a system matrix is constructed as 
```math
K_{ij} = \frac{\partial r_i}{\partial a_i}
```
Each entry, $r_i$, in $\boldsymbol{r}$, typically depend on only a few of the unknowns, $a_j$, such that many entries in $K_{ij}$ will be zero. This so-called sparse structure of the linear system, $K_{ij}$, depends on many factors such as e.g. the weak ...

----------

With the current text, it is for example not obvious that $K_{ij}$ can be zero while $K_{ji}$ is not. (I.e. which is nonzero "DoFs `i` and `j` couple"). 

(Not sure if it makes sense to discuss cases when using alternative techniques for defining the stiffness matrix via quasi-newton methods here, perhaps just mention that when discussing the `coupling` keyword?)

Reviewer: This point also picked my attention. I would not even to down to nonlinear problems here. The point is really about linear operators, which either appear naturally as part of a PDE description (see e.g. Navier-Stokes example or Poisson problem or even linear elasticity) or in the linearization (e.g. second variation in energy-based elasticity). Hence the sparse matrix is not directly about linear systems, but about the representation of the discretization of an operator which allows e.g. efficient linear solves and fast simulations. So I have something into this direction in my head:
```suggestion
The sparse structure of linear operators in the context of the simulation of partial differential equations depends on many factors such as e.g. the weak
```
but it needs further refinement. Does this make sense for you guys?

   `(2, 3)` included in the sparsity pattern)
 - Trial function 3 couples with test functions 2, 3, and 4 (entries `(3, 2)`, `(3, 3)`, and
   `(3, 4)` included in the sparsity pattern)
 - Trial function 4 couples with test functions 3 and 4 (entries `(4, 3)` and `(4, 4)`
   included in the sparsity pattern)

The resulting sparsity pattern would look like this:

```
4×4 SparseArrays.SparseMatrixCSC{Float64, Int64} with 10 stored entries:
 0.0  0.0   ⋅    ⋅ 
 0.0  0.0  0.0   ⋅ 
  ⋅   0.0  0.0  0.0
  ⋅    ⋅   0.0  0.0
```

Moreover, if the problem is solved with periodic boundary conditions, for example by constraining the
value on the right side to the value on the left side, there will be additional couplings.
In the example above, this means that DoF 4 should be equal to DoF 1. Since DoF 4 is
constrained it has to be eliminated from the system. Existing entries that include DoF 4 are
`(3, 4)`, `(4, 3)`, and `(4, 4)`. Given the simple constraint in this case we can simply
replace DoF 4 with DoF 1 in these entries and we end up with entries `(3, 1)`, `(1, 3)`, and
`(1, 1)`. This results in two new entries: `(3, 1)` and `(1, 3)` (entry `(1, 1)` is already

> included).

Reviewer 3: While this is a very nice introductory text, I think it hides the important things below.
Perhaps add at least the example in a collapsible block?

## Creating sparsity patterns

Creating a sparsity pattern can be quite expensive if not done properly and therefore
Ferrite provides efficient methods and datastructures for this. In general the sparsity
pattern is not known in advance and has to be created incrementally. To make this
incremental construction efficient it is necessary to use a dynamic data structure which
allow for fast insertions.

The sparsity pattern also serves as a "matrix builder". When all entries are inserted into
the sparsity pattern the dynamic data structure is typically converted, or "compressed",
into a sparse matrix format such as e.g. the [*compressed sparse row
(CSR)*](https://en.wikipedia.org/wiki/Sparse_matrix#Compressed_sparse_row_(CSR,_CRS_or_Yale_format))
format or the [*compressed sparse column
(CSC)*](https://en.wikipedia.org/wiki/Sparse_matrix#Compressed_sparse_column_(CSC_or_CCS))

> format, which is the default sparse matrix type implemented in the SparseArrays standard

Reviewer 3: ```suggestion
format, where the latter is the default sparse matrix type implemented in the SparseArrays standard
```

library. These matrix formats allow for fast linear algebra operations, such as
factorizations and matrix-vector multiplications, that are needed when the linear system is
solved. See [Instantiating the sparse matrix](@ref) for more details.

In summary, a dynamic structure is more efficient when incrementally building the pattern by
inserting new entries, and a static or compressed structure is more efficient for linear
algebra operations.

### Basic sparsity patterns construction in Ferrite

Working with the sparsity pattern explicitly is in many cases not necessary. For basic
usage (e.g. when only one matrix needed, when no customization of the pattern is
required, etc) there exist convenience methods of [`create_matrix`](@ref) that return
the matrix directly. For example, most examples in this documentation don't deal with
the sparsity pattern explicitly.
    

> ### Costum sparsity patterns construction in Ferrite

Reviewer 3: ```suggestion
### Custom sparsity patterns construction in Ferrite
```

In more advanced cases, we need more fine grained control of the sparsity pattern. The 
following steps are typically taken when constructing a sparsity pattern in Ferrite:

>  1. *Initialize an empty pattern.* This can be done by either using the

Reviewer 3: ```suggestion
 1. **Initialize an empty pattern:** This can be done by either using the
```

    [`init_sparsity_pattern(dh)`](@ref) function or by using a constructor directly.
    `init_sparsity_pattern` will return a default pattern type that is compatible with the
    DofHandler. In some cases you might require another type of pattern (for example a

>     blocked pattern) and in that case you can use the constructor directly.

Reviewer 3: ```suggestion
    blocked pattern, see [Blocked sparsity pattern](@ref)) and in that case you can use the constructor directly.
```

>  2. *Add entries to the pattern.* There are a number of functions that add entries to the

Reviewer 3: ```suggestion
 2. **Add entries to the pattern:** There are a number of functions that add entries to the
```

>  2. *Add entries to the pattern.* There are a number of functions that add entries to the
>     pattern:
>      - [`create_sparsity_pattern!`](@ref) adds entries for all couplings between the DoFs
>        within each element. These entries correspond to assembling the standard element
>        matrix and is thus almost always required.
>      - [`cross_element_coupling!`](@ref) adds entries for couplings between the DoFs in
>        neighboring elements. These entries are required when integrating along internal
>        interfaces between elements (e.g. for discontinuous Galerkin methods).
>      - [`Ferrite.add_entry!`](@ref) adds a single entry to the pattern. This can be used if
>        you need to add custom entries that are not covered by the other functions.
> 
>  3. *Condense the pattern.* This is done with the [`condense_sparsity_pattern!`](@ref)
>     function and adds entries required from constraints and boundary conditions in the
>     ConstraintHandler. Note that condensing the pattern *must* be done as the last operation
>     on the pattern, e.g. after adding all other entries.

Reviewer 3: As discussed on Slack...
```suggestion
 2. **Add entries to the pattern:** This can conveniently be done by using [`create_sparsity_pattern!`](@ref). If a finer-grained control is required, there are a number of functions that add entries to the
    pattern:
     - [`cell_couplings!`](@ref) adds entries for all couplings between the DoFs
       within each element. These entries correspond to assembling the standard element
       matrix and is thus almost always required.
     - [`cross_element_coupling!`](@ref) adds entries for couplings between the DoFs in
       neighboring elements. These entries are required when integrating along internal
       interfaces between elements (e.g. for discontinuous Galerkin methods).
     - [`Ferrite.add_entry!`](@ref) adds a single entry to the pattern. This can be used if
       you need to add custom entries that are not covered by the other functions.
     - [`condense_sparsity_pattern!`](@ref) adds entries required from constraints and boundary conditions in the
    ConstraintHandler. Note that condensing the pattern *must* be done as the last operation
    on the pattern, e.g. after adding all other entries. 
```

Reviewer 3: With the extra comments that I think `constraint_couplings!` is a better name than `condense_sparsity_pattern!`. 

Side-note: Sounds like having to call `condense_sparsity_pattern!` last can easily lead to user bugs. Would it make sense to add a `condensed::Ref{Bool}` (or `ScalarWrapper`) in `SparsityPattern` to allow the other methods error if trying to add entries after condensation?

Reviewer 4: If you dont call the condense function last, I think you will at least get an error in the assembly-routine.

I think the name change make sense!

Reviewer: > With the extra comments that I think `constraint_couplings!` is a better name than `condense_sparsity_pattern!`.
> 

The process is called condensation, right? So why should we use a different name for this?

> Side-note: Sounds like having to call `condense_sparsity_pattern!` last can easily lead to user bugs. Would it make sense to add a `condensed::Ref{Bool}` (or `ScalarWrapper`) in `SparsityPattern` to allow the other methods error if trying to add entries after condensation?

I think this is a really good idea.

Reviewer 3: > The process is called condensation, right? So why should we use a different name for this?

Sorry, missed this comment @termi-official: My logic is that the condensation happens when you call `apply!`, but this step only adds additional entries due to the couplings introduced by the affine constraints.

Reviewer: No problem. I see. I still think users reading `constraint_couplings` would expect something different happening. What about `add_constraint_pattern!` or even something like this?

Reviewer 3: Yes, probably good to be consistent with using `add_XY!` or `XY!` here, and since 
`add_entry!` is kindof needed perhaps having add makes sense. For the `couplings` part, was just to be consistent with `cross_element_coupling!`. 
So the options could be
* `add_cell_pattern!`
* `add_cross_element_pattern!` (cell?)
* `add_entry!`
* `add_constraint_pattern!`
or 
* `add_cell_couplings!`
* `add_cross_element_couplings!` (cell?)
* `add_entry!`
* `add_constraint_couplings!`

I think I prefer `couplings`, as this is more logical for the others IMO.

Reviewer: For me `pattern` makes sense too for the suffix since we insert some kind of pattern to the sparsity pattern.

What about `cross_element` -> `interface`?

Author: Perhaps
```julia
insert_cell_couplings!(...)
insert_interface_couplings!(...)
insert_constraint_couplings!(...)
insert_entry!(...)
```
or
```julia
create_cell_couplings!(...)
create_interface_couplings!(...)
create_constraint_couplings!(...)
create_coupling!(...)
```
or
```julia
add_cell_couplings!(...)
add_interface_couplings!(...)
add_constraint_couplings!(...)
add_coupling!(...)
```

Author: A typo I think, but any of the suggestions could work with `_coupling(s)` -> `entries` too.

>  4. *Instantiate the matrix*. A sparse matrix can be created from the sparsity pattern using

Reviewer 3: ```suggestion
 4. **Instantiate the matrix:** A sparse matrix can be created from the sparsity pattern using
```

    [`create_matrix`](@ref), see [Instantiating the sparse matrix](@ref) below for more
    details.

!!! note "Swiss army knife"
    The [`create_sparsity_pattern!`](@ref) function is something of a swiss army knife since
    it can do more than just adding the basic element couplings as described above. For
    example, when passed the correct arguments, it can also do the job of
    `cross_element_coupling!` and `condense_sparsity_pattern!`. Refer to the API reference

>     for details.

Reviewer 3: To be removed if adopting my suggestion with `cell_couplings!` above

### Increasing the sparsity

By default when creating a sparsity pattern it is assumed that a DoF couple with all other
DoFs in the same element.

is created such that DoFs couple with all DoFs in the same
element. 

This is the most general case and results in a dense
matrix. However, in many cases the sparsity can be increased by reducing the number of DoFs
that couple with each other. This can be done by using the `coupling` keyword argument in
the `create_sparsity_pattern` function. This keyword argument takes a function that returns
a boolean value indicating if two DoFs should couple or not. The function should take two
arguments, the first being the element index and the second being the DoF index within the

> element. The function should return `true` if the DoFs should couple and `false` otherwise.

Reviewer 3: This sentence doesn't make sense to me. 
I would have expected that `f(i, j)` should say if DoF `i` couples to DoF `j`? But here it sounds like the `cellid` is expected, but perhaps I'm just confused with array element vs finite element?

Wasn't there an option to give the field coupling via a `Matrix{Bool}` before?

Side-note: Probably giving `uncoupled::Vector{Pair{Symbol,Symbol}}` would be more user-friendly to remove couplings instead, but can be done later...

Author: This paragraph is Copilot hallucinations I forgot to remove haha

Discuss the `coupling` keyword argument.

Discuss the `keep_constrained` keyword argument.

## Creating the sparsity pattern

Given a `DofHandler` we can obtain the corresponding sparse matrix by using the

> [`create_sparsity_pattern`](@ref) function. This will setup a `SparseMatrixCSC`

Reviewer 3: ```suggestion
[`create_matrix`](@ref) function. This will setup a `SparseMatrixCSC`
```
?

# Sparsity pattern and sparse matrix

## Sparsity Pattern

Given a `DofHandler` we can obtain the corresponding sparse matrix by using the
[`create_sparsity_pattern`](@ref) function. This will setup a `SparseMatrixCSC`
with stored values on all the places corresponding to the degree of freedom numbering
in the `DofHandler`. This means that when we assemble into the global stiffness matrix
there is no need to change the internal representation of the sparse matrix since the
sparse structure will not change.

Often the finite element problem is symmetric and will result in a symmetric sparse
matrix. This information is often something that the sparse solver can take advantage of.
If the solver only needs half the matrix there is no need to assemble both halves.
For this purpose there is a [`create_symmetric_sparsity_pattern`](@ref) function that
will only create the upper half of the matrix, and return a `Symmetric` wrapped
`SparseMatrixCSC`.

Given a `DofHandler` `dh` we can obtain the (symmetric) sparsity pattern as

```julia
K = create_matrix(dh)

> K = create_symmetric_sparsity_pattern(dh)

Reviewer: I think this is not correct?

Author: Yea, missed this. Actually I removed symmetric support on the pattern level, but this is one point to discuss.

Reviewer 3: Unresolved this since this looks inconsistent here, shouldn't it at least be `create_symmetric_matrix(dh)`, or removed if not used anymore?

 4. *Instantiate the matrix*. A sparse matrix can be created from the sparsity pattern using
    [`create_matrix`](@ref), see [Instantiating the sparse matrix](@ref) below for more
    details.

!!! note "Swiss army knife"
    The [`create_sparsity_pattern!`](@ref) function is something of a swiss army knife since
    it can do more than just adding the basic element couplings as described above. For
    example, when passed the correct arguments, it can also do the job of
    `cross_element_coupling!` and `condense_sparsity_pattern!`. Refer to the API reference
    for details.

!!! note "Advanced usage"
    Working with the sparsity pattern explicitly is in many cases not necessary. For basic
    usage (e.g. when only one matrix needed, when no customization of the pattern is
    required, etc) there exist convenience methods of [`create_matrix`](@ref) that return
    the matrix directly. For example, most examples in this documentation don't deal with
    the sparsity pattern explicitly.

### Increasing the sparsity

By default when creating a sparsity pattern it is assumed that a DoF couple with all other
DoFs in the same element.

> is created such that DoFs couple with all DoFs in the same

Reviewer 2: Remove?
